server/src/com/thoughtworks/go/server/controller/AgentRegistrationController.java

```diff
@@ -192,36 +191,27 @@ public ModelAndView agentRequest(@RequestParam("hostname") String hostname,
         final String ipAddress = request.getRemoteAddr();
         LOG.debug("Processing registration request from agent [{}/{}]", hostname, ipAddress);
         Registration keyEntry;
-        String preferredHostname = hostname;
 
         try {
-            if (goConfigService.serverConfig().shouldAutoRegisterAgentWith(agentAutoRegisterKey)) {
-                preferredHostname = getPreferredHostname(agentAutoRegisterHostname, hostname);
-                LOG.info("[Agent Auto Registration] Auto registering agent with uuid {} ", uuid);
-            } else {
-                if (elasticAgentAutoregistrationInfoPresent(elasticAgentId, elasticPluginId)) {
-                    throw new RuntimeException(String.format("Elastic agent registration requires an auto-register agent key to be setup on the server. Agent-id: [%s], Plugin-id: [%s]", elasticAgentId, elasticPluginId));
-                }
-            }
+            final boolean shouldAutoRegisterAgent = goConfigService.serverConfig().shouldAutoRegisterAgentWith(agentAutoRegisterKey);
+            final boolean registeredAlready = goConfigService.hasAgent(uuid);
```

Comment:
Ok, there is one slight problem with moving this code on top. 

If you look at the existing code, when auto-registration properties file is setup correctly for an agent, [that](https://github.com/gocd/gocd/blob/da55ed3cf9ce35641e96f4f18a8787fced3199bb/server/src/com/thoughtworks/go/server/controller/AgentRegistrationController.java#L226) line will take care of saving the agent config to the file, so when the call to check if the agent was registered already happens [here](https://github.com/gocd/gocd/blob/da55ed3cf9ce35641e96f4f18a8787fced3199bb/server/src/com/thoughtworks/go/server/controller/AgentRegistrationController.java#L238), it would return `true`. This causes `AgentRuntimeInfo.fromServer` to set the agent status to `Idle` which would then cause a successful registration of the agent in `agentService.requestRegistration`. 

With this PR, [that](https://github.com/bdpiparva/gocd/blob/d9c9006a9f9c2713aaf0e6cc2a9e315503df7c0b/server/src/com/thoughtworks/go/server/controller/AgentRegistrationController.java#L197) line will return `false` for such an agent causing `agentService.requestRegistration` use a different flow. You might have to recheck for the value of the flag after the auto-registration but runs. 



 correctly, [that](https://github.com/bdpiparva/gocd/blob/d9c9006a9f9c2713aaf0e6cc2a9e315503df7c0b/server/src/com/thoughtworks/go/server/controller/AgentRegistrationController.java#L216) line will take care of saving the agent in the config, and hence the `registeredAlready` flag